astropy/time/tests/test_precision.py

```diff
@@ -142,69 +293,376 @@ def test_leap_seconds_rounded_correctly():
-    """Conversion can leave jd1 not an integer"""
-    scale1 = 'tai'
-    scale2 = 'tcb'
-    jd1, jd2 = 0., 0.
+@given(jd_arrays(unreasonable_jd()))
+def test_day_frac_always_less_than_half(jds):
+    jd1, jd2 = jds
+    t_jd1, t_jd2 = day_frac(jd1, jd2)
+    assert np.all(t_jd1 % 1 == 0)
+    assert np.all(abs(t_jd2) <= 0.5)
+    assert np.all((abs(t_jd2) < 0.5) | (t_jd1 % 2 == 0))
+
+
+@given(integers(-10**8, 10**8), sampled_from([-0.5, 0.5]))
+def test_day_frac_round_to_even(jd1, jd2):
+    t_jd1, t_jd2 = day_frac(jd1, jd2)
+    assert (abs(t_jd2) == 0.5) and (t_jd1 % 2 == 0)
+
+
+@given(scale=sampled_from(STANDARD_TIME_SCALES), jds=unreasonable_jd())
+@example(scale="tai", jds=(0.0, 0.0))
+@example(scale="tai", jds=(0.0, -31738.500000000346))
+def test_resolution_never_decreases(scale, jds):
+    jd1, jd2 = jds
+    assume(not scale == 'utc' or 2440000 < jd1 + jd2 < 2460000)
+    t = Time(jd1, jd2, format="jd", scale=scale)
+    with quiet_erfa():
+        assert t != t + dt_tiny
+
+
+@given(reasonable_jd())
+def test_resolution_never_decreases_utc(jds):
+    """UTC is very unhappy with unreasonable times"""
+    jd1, jd2 = jds
+    t = Time(jd1, jd2, format="jd", scale="utc")
+    with quiet_erfa():
+        assert t != t + dt_tiny
+
+
+@given(scale1=sampled_from(STANDARD_TIME_SCALES),
+       scale2=sampled_from(STANDARD_TIME_SCALES),
+       jds=unreasonable_jd())
+@example(scale1='tcg', scale2='ut1', jds=(2445149.5, 0.47187700984387526))
+@example(scale1='tai', scale2='tcb', jds=(2441316.5, 0.0))
+@example(scale1='tai', scale2='tcb', jds=(0.0, 0.0))
+def test_conversion_preserves_jd1_jd2_invariant(iers_b, scale1, scale2, jds):
+    jd1, jd2 = jds
     t = Time(jd1, jd2, scale=scale1, format="jd")
-    with pytest.warns(
-            ErfaWarning,
-            match=r'ERFA function "taiutc" yielded 1 of "dubious year'):
-        t2 = getattr(t, scale2)
+    try:
+        with quiet_erfa():
+            t2 = getattr(t, scale2)
+    except iers.IERSRangeError:  # UT1 conversion needs IERS data
+        assume(False)
+    except ErfaError:
+        assume(False)
     assert t2.jd1 % 1 == 0
     assert abs(t2.jd2) <= 0.5
     assert abs(t2.jd2) < 0.5 or t2.jd1 % 2 == 0
 
 
-def test_conversion_preserves_jd1_jd2_invariant_2():
-    """Conversion can leave abs(jd2)>0.5"""
-    scale1 = 'tai'
-    scale2 = 'tcb'
-    jd1, jd2 = (2441316.5, 0.0)
+@given(scale1=sampled_from(STANDARD_TIME_SCALES),
+       scale2=sampled_from(STANDARD_TIME_SCALES),
+       jds=unreasonable_jd())
+def test_conversion_never_loses_precision(iers_b, scale1, scale2, jds):
+    jd1, jd2 = jds
     t = Time(jd1, jd2, scale=scale1, format="jd")
-    t2 = getattr(t, scale2)
-    assert t2.jd1 % 1 == 0
-    assert abs(t2.jd2) <= 0.5
-    assert abs(t2.jd2) < 0.5 or t2.jd1 % 2 == 0
+    try:
+        # If the rates ever differ, might lose one ULP
+        with quiet_erfa():
+            t2 = t + 2*dt_tiny
+            assert getattr(t, scale2) < getattr(t2, scale2)
```

Comment:
So, I guess somehow with the oldest dependencies an extra ULP was lost, and it is not surprising this is for UTC since that involves converting the time scale to TAI and back. My sense would be to add `3*dt_tiny` if both scales are UTC.

Reply:
I'm not *entirely* sure why this happens, but after a few attempts I gave up and slapped an "xfail under Numpy 1.16" decorator on it.

Reply:
Hmm, looks like decorator is on wrong function?  I'll fix...

Reply:
:man_facepalming: haha, thanks...